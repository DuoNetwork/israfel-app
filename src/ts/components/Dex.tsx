import { Layout } from 'antd';
import { Affix } from 'antd';
import * as React from 'react';
// import { IWsUserOrderResponse } from '../common/types';
import { SDivFlexCenter } from './_styled';
import OperationCard from './Cards/OperationCard';
// import OperationHistory from './Cards/OperationHistory';
// import OperationCard from './Cards/OperationCard';
// import OrderbookCardSubscription from './Cards/OrderbookCardSubscription';
import Header from './Header';

export interface IProps {
	// wsUserOrderResponse: IWsUserOrderResponse;
	network: number;
	locale: string;
	account: string;
	updateLocale: (locale: string) => any;
	// refresh: () => any;
	// subscription: (marketId: string, pair: string) => any;
	// submitOrders: (amount: number, price: number, action: string) => any;
}

export default class Dex extends React.PureComponent<IProps> {
	public render() {
		const { locale, network, updateLocale } = this.props;
		return (
			<Layout>
				<div className="App">
					<Header locale={locale} network={network} updateLocale={updateLocale} />
					<SDivFlexCenter center horizontal>
						{/* <OperationHistory
							askBidMsg={wsUserOrderResponse.userOrder}
							locale={locale}
						/> */}
						<Affix offsetTop={10}>
							<OperationCard />
						</Affix>
					</SDivFlexCenter>
				</div>
			</Layout>
		);
	}
}
