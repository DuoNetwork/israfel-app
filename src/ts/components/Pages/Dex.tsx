import { Spin } from 'antd';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import {
	IAcceptedPrice,
	ICustodianInfo,
	IEthBalance,
	INotification,
	IOrderBookSnapshot,
	IToken,
	ITokenBalance,
	IUserOrder
} from 'ts/common/types';
import util from 'ts/common/util';
import { SDivFlexCenter } from '../_styled';
import BalanceCard from '../Cards/BalanceCard';
import ConvertCard from '../Cards/ConvertCard';
import CustodianCard from '../Cards/CustodianCard';
import OrderHistoryCard from '../Cards/OrderHistoryCard';
import TradeCard from '../Cards/TradeCard';

interface IProps {
	account: string;
	ethBalance: IEthBalance;
	tokens: IToken[];
	acceptedPrices: { [custodian: string]: IAcceptedPrice[] };
	ethPrice: number;
	custodians: { [custodian: string]: ICustodianInfo };
	custodianTokenBalances: { [custodian: string]: { [code: string]: ITokenBalance } };
	orderBooks: { [pair: string]: IOrderBookSnapshot };
	orderHistory: { [pair: string]: IUserOrder[] };
	connection: boolean;
	notify: (notification: INotification) => any;
	subscribeOrder: (account: string) => any;
	unsubscribeOrder: () => any;
}

interface IState {
	account: string;
	convertCustodian: string;
	convertAToken: string;
	convertBToken: string;
	tradeToken: string;
	showBalances: boolean;
}

export default class Dex extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			account: props.account,
			convertCustodian: '',
			convertAToken: '',
			convertBToken: '',
			tradeToken: '',
			showBalances: true
		};
	}

	public componentDidMount() {
		document.title = 'DUO | Trustless Derivatives';
		this.props.subscribeOrder(this.props.account);
	}

	public static getDerivedStateFromProps(props: IProps, state: IState) {
		if (props.account !== state.account) {
			props.subscribeOrder(props.account);
			return {
				account: props.account
			};
		}

		return null;
	}

	public componentWillUnmount() {
		this.props.unsubscribeOrder();
	}

	public handleConvert = (custodian: string, aToken: string, bToken: string) =>
		this.setState({
			convertCustodian: custodian.toLocaleLowerCase(),
			convertAToken: aToken,
			convertBToken: bToken
		});

	public handleTrade = (token: string) => this.setState({ tradeToken: token });

	public render() {
		const {
			account,
			tokens,
			acceptedPrices,
			custodians,
			custodianTokenBalances,
			ethBalance,
			connection,
			orderBooks,
			ethPrice,
			orderHistory,
			notify
		} = this.props;
		const {
			convertCustodian,
			tradeToken,
			convertAToken,
			convertBToken,
			showBalances
		} = this.state;
		const beethovenList: string[] = [];
		const mozartList: string[] = [];
		for (const custodian in custodians) {
			const info = custodians[custodian];
			const code = info.code.toLowerCase();
			if (code.startsWith(CST.BEETHOVEN.toLowerCase())) beethovenList.push(custodian);
			else if (code.startsWith(CST.MOZART.toLowerCase())) mozartList.push(custodian);
		}
		beethovenList.sort((a, b) => custodians[a].states.maturity - custodians[b].states.maturity);
		mozartList.sort((a, b) => custodians[a].states.maturity - custodians[b].states.maturity);
		const tradeTokenInfo = tokens.find(t => t.code === tradeToken);
		const tradeTokenBalance = tradeTokenInfo
			? custodianTokenBalances[tradeTokenInfo.custodian][tradeToken]
			: undefined;
		const tokenInterestOrLeverage =
			tradeTokenInfo &&
			acceptedPrices[tradeTokenInfo.custodian] &&
			acceptedPrices[tradeTokenInfo.custodian].length
				? util.getTokenInterestOrLeverage(
						custodians[tradeTokenInfo.custodian],
						acceptedPrices[tradeTokenInfo.custodian][
							acceptedPrices[tradeTokenInfo.custodian].length - 1
						],
						tradeToken.startsWith('a') || tradeToken.startsWith('s')
				)
				: 0;
		let tokenNavInEth = 0;
		let tokenNavUpdatedAt = 0;
		if (tradeTokenInfo && custodians[tradeTokenInfo.custodian]) {
			if (tradeToken.startsWith('a') || tradeToken.startsWith('s'))
				tokenNavInEth = custodians[tradeTokenInfo.custodian].states.navA / ethPrice;
			else tokenNavInEth = custodians[tradeTokenInfo.custodian].states.navB / ethPrice;

			tokenNavUpdatedAt = custodians[tradeTokenInfo.custodian].states.lastPriceTime;
		}

		return (
			<div>
				<Spin spinning={!connection} tip="loading...">
					<SDivFlexCenter
						center
						horizontal
						marginBottom="20px"
						style={{ paddingTop: '20px' }}
					>
						{beethovenList.map(c => {
							const tbs = custodianTokenBalances[c] || {};
							const obs: { [pair: string]: IOrderBookSnapshot } = {};
							for (const code in tbs) {
								const pair = code + '|' + CST.TH_WETH;
								obs[pair] = orderBooks[pair];
							}
							return (
								<CustodianCard
									key={c}
									type={CST.BEETHOVEN}
									custodian={c}
									handleConvert={this.handleConvert}
									handleTrade={this.handleTrade}
									info={custodians[c]}
									margin="0 10px"
									acceptedPrices={acceptedPrices[c]}
									tokenBalances={tbs}
									orderBooks={obs}
									ethPrice={ethPrice}
								/>
							);
						})}
					</SDivFlexCenter>
					<SDivFlexCenter center horizontal marginBottom="20px">
						{mozartList.map(c => {
							const tbs = custodianTokenBalances[c] || {};
							const obs: { [pair: string]: IOrderBookSnapshot } = {};
							for (const code in tbs) {
								const pair = code + '|' + CST.TH_WETH;
								obs[pair] = orderBooks[pair];
							}
							return (
								<CustodianCard
									key={c}
									type={CST.MOZART}
									custodian={c}
									handleConvert={this.handleConvert}
									handleTrade={this.handleTrade}
									info={custodians[c]}
									margin="0 10px"
									acceptedPrices={acceptedPrices[c]}
									tokenBalances={tbs}
									orderBooks={obs}
									ethPrice={ethPrice}
								/>
							);
						})}
					</SDivFlexCenter>
					<OrderHistoryCard orderHistory={orderHistory} account={account} />
					<ConvertCard
						account={account}
						tokenBalances={custodianTokenBalances[convertCustodian]}
						ethBalance={ethBalance}
						custodian={convertCustodian}
						aToken={convertAToken}
						bToken={convertBToken}
						info={custodians[convertCustodian]}
						notify={notify}
						handleClose={() => this.handleConvert('', '', '')}
					/>
					<TradeCard
						account={account}
						token={tradeToken}
						tokenInfo={tradeTokenInfo}
						tokenBalance={tradeTokenBalance}
						ethBalance={ethBalance}
						orderBook={
							orderBooks[tradeToken + '|' + CST.TH_WETH] || {
								pair: 'pair',
								version: 0,
								bids: [],
								asks: []
							}
						}
						ethPrice={ethPrice}
						navInEth={tokenNavInEth}
						navUpdatedAt={tokenNavUpdatedAt}
						notify={notify}
						interestOrLeverage={tokenInterestOrLeverage}
						handleClose={() => this.handleTrade('')}
					/>
				</Spin>
				<BalanceCard
					ethPrice={ethPrice}
					visible={showBalances}
					account={account}
					ethBalance={ethBalance}
					beethovenList={beethovenList}
					mozartList={mozartList}
					custodians={custodians}
					custodianTokenBalances={custodianTokenBalances}
					notify={notify}
					handleClose={() => this.setState({ showBalances: !showBalances })}
				/>
			</div>
		);
	}
}
