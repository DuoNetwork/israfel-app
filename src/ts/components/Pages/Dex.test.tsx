/**
 * @jest-environment jsdom
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import Dex from './Dex';

describe('Dex Test', () => {
	describe('User Login', () => {
		it('Test Snapshot', () => {
			const account = 'test';
			const ethBalance = {
				allowance: 1.157920892373162,
				eth: 0.5076699707487066,
				weth: 0.984992
			};
			const acceptedPrices = {
				'0x00be45fe5903ab1b33a9d3969b05b29552a6d18b': [{
					blockNumber: 9835631,
					contractAddress: '0x00be45Fe5903AB1b33a9d3969b05b29552a6d18b',
					navA: 0.9554321303002988,
					navB: 1.0891357393994023,
					price: 128.45782891277358,
					timestamp: 1545782400000,
					transactionHash:
						'0x54c64c781b92e1d44d5229245e86b9aae602f126339e16d72f50162100e7b5e5'
				}]
			};
			const ethPrice = 123;
			const custodians = {
				'0x00be45fe5903ab1b33a9d3969b05b29552a6d18b': {
					code: 'MOZART-PPT',
					states: {
						alpha: 0.5,
						beta: 1,
						createCommRate: 0.01,
						ethCollateral: 12.358884504282397,
						feeBalance: 0.15089954869517602,
						iterationGasThreshold: 65000,
						lastOperationTime: 0,
						lastPrice: 145.11564352568783,
						lastPriceTime: 1546423200000,
						limitLower: 0.25,
						limitPeriodic: 0,
						limitUpper: 1.75,
						maturity: 0,
						minBalance: 0.01,
						navA: 0.8199774206003582,
						navB: 1.3600451587992837,
						nextResetAddrIndex: 0,
						operationCoolDown: 86400000,
						period: 3600000,
						periodCoupon: 0,
						preResetWaitingBlocks: 10,
						priceFetchCoolDown: 3000000,
						redeemCommRate: 0.01,
						resetPrice: 122.97700574468506,
						resetPriceTime: 1545530400000,
						resetState: 'UpwardReset',
						state: 'Trading',
						totalSupplyA: 506.61953689367846,
						totalSupplyB: 1013.2390737873569,
						totalUsers: 18
					}
				}
			};
			const custodianTokenBalances = {
				'0x00be45fe5903ab1b33a9d3969b05b29552a6d18b': {
					LETH: { balance: 0, allowance: 0 },
					sETH: { balance: 0, allowance: 1.157920892373162e59 }
				}
			};
			const orderBooks = {
				'LETH-M19|WETH': {
					asks: [],
					bids: [],
					pair: 'LETH-M19|WETH',
					version: 1546424305772
				}
			};
			const orderHistory = {};
			const connection = true;
			const notify = jest.fn(() => ({}));
			const subscribeOrder = jest.fn();
			const unsubscribeOrder = jest.fn();
			const wrapper = shallow(
				<Dex
					orderHistory={orderHistory}
					connection={connection}
					notify={notify}
					subscribeOrder={subscribeOrder}
					unsubscribeOrder={unsubscribeOrder}
					custodianTokenBalances={custodianTokenBalances}
					custodians={custodians}
					orderBooks={orderBooks}
					ethPrice={ethPrice}
					acceptedPrices={acceptedPrices}
					ethBalance={ethBalance}
					account={account}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			});
	});
});
