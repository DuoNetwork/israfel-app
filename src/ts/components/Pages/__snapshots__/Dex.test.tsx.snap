// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dex Test User Login Test Snapshot 1`] = `
<div>
  <Spin
    prefixCls="ant-spin"
    size="default"
    spinning={false}
    tip="loading..."
    wrapperClassName=""
  >
    <styled.div
      center={true}
      horizontal={true}
      marginBottom="20px"
      style={
        Object {
          "paddingTop": "20px",
        }
      }
    >
      <CustodianCard
        key="0"
        margin="0 10px"
        type="Beethoven"
      />
      <CustodianCard
        key="1"
        margin="0 10px"
        type="Beethoven"
      />
    </styled.div>
    <styled.div
      center={true}
      horizontal={true}
      marginBottom="20px"
    >
      <CustodianCard
        acceptedPrices={
          Array [
            Object {
              "blockNumber": 9835631,
              "contractAddress": "0x00be45Fe5903AB1b33a9d3969b05b29552a6d18b",
              "navA": 0.9554321303002988,
              "navB": 1.0891357393994023,
              "price": 128.45782891277358,
              "timestamp": 1545782400000,
              "transactionHash": "0x54c64c781b92e1d44d5229245e86b9aae602f126339e16d72f50162100e7b5e5",
            },
          ]
        }
        custodian="0x00be45fe5903ab1b33a9d3969b05b29552a6d18b"
        ethPrice={123}
        handleConvert={[Function]}
        handleTrade={[Function]}
        info={
          Object {
            "code": "MOZART-PPT",
            "states": Object {
              "alpha": 0.5,
              "beta": 1,
              "createCommRate": 0.01,
              "ethCollateral": 12.358884504282397,
              "feeBalance": 0.15089954869517602,
              "iterationGasThreshold": 65000,
              "lastOperationTime": 0,
              "lastPrice": 145.11564352568783,
              "lastPriceTime": 1546423200000,
              "limitLower": 0.25,
              "limitPeriodic": 0,
              "limitUpper": 1.75,
              "maturity": 0,
              "minBalance": 0.01,
              "navA": 0.8199774206003582,
              "navB": 1.3600451587992837,
              "nextResetAddrIndex": 0,
              "operationCoolDown": 86400000,
              "period": 3600000,
              "periodCoupon": 0,
              "preResetWaitingBlocks": 10,
              "priceFetchCoolDown": 3000000,
              "redeemCommRate": 0.01,
              "resetPrice": 122.97700574468506,
              "resetPriceTime": 1545530400000,
              "resetState": "UpwardReset",
              "state": "Trading",
              "totalSupplyA": 506.61953689367846,
              "totalSupplyB": 1013.2390737873569,
              "totalUsers": 18,
            },
          }
        }
        key="0x00be45fe5903ab1b33a9d3969b05b29552a6d18b"
        margin="0 10px"
        orderBooks={
          Object {
            "LETH|WETH": undefined,
            "sETH|WETH": undefined,
          }
        }
        tokenBalances={
          Object {
            "LETH": Object {
              "allowance": 0,
              "balance": 0,
            },
            "sETH": Object {
              "allowance": 1.157920892373162e+59,
              "balance": 0,
            },
          }
        }
        type="Mozart"
      />
      <CustodianCard
        key="0"
        margin="0 10px"
        type="Mozart"
      />
    </styled.div>
     
    <OrderHistoryCard
      account="test"
      orderHistory={Object {}}
    />
    <ConvertCard
      aToken=""
      account="test"
      bToken=""
      custodian=""
      ethBalance={
        Object {
          "allowance": 1.157920892373162,
          "eth": 0.5076699707487066,
          "weth": 0.984992,
        }
      }
      handleClose={[Function]}
      notify={[MockFunction]}
    />
    <TradeCard
      account="test"
      ethBalance={
        Object {
          "allowance": 1.157920892373162,
          "eth": 0.5076699707487066,
          "weth": 0.984992,
        }
      }
      ethPrice={123}
      handleClose={[Function]}
      interestOrLeverage={0}
      navInEth={0}
      navUpdatedAt={0}
      notify={[MockFunction]}
      orderBook={
        Object {
          "asks": Array [],
          "bids": Array [],
          "pair": "pair",
          "version": 0,
        }
      }
      token=""
    />
  </Spin>
  <BalanceCard
    account="test"
    beethovenList={Array []}
    custodianTokenBalances={
      Object {
        "0x00be45fe5903ab1b33a9d3969b05b29552a6d18b": Object {
          "LETH": Object {
            "allowance": 0,
            "balance": 0,
          },
          "sETH": Object {
            "allowance": 1.157920892373162e+59,
            "balance": 0,
          },
        },
      }
    }
    custodians={
      Object {
        "0x00be45fe5903ab1b33a9d3969b05b29552a6d18b": Object {
          "code": "MOZART-PPT",
          "states": Object {
            "alpha": 0.5,
            "beta": 1,
            "createCommRate": 0.01,
            "ethCollateral": 12.358884504282397,
            "feeBalance": 0.15089954869517602,
            "iterationGasThreshold": 65000,
            "lastOperationTime": 0,
            "lastPrice": 145.11564352568783,
            "lastPriceTime": 1546423200000,
            "limitLower": 0.25,
            "limitPeriodic": 0,
            "limitUpper": 1.75,
            "maturity": 0,
            "minBalance": 0.01,
            "navA": 0.8199774206003582,
            "navB": 1.3600451587992837,
            "nextResetAddrIndex": 0,
            "operationCoolDown": 86400000,
            "period": 3600000,
            "periodCoupon": 0,
            "preResetWaitingBlocks": 10,
            "priceFetchCoolDown": 3000000,
            "redeemCommRate": 0.01,
            "resetPrice": 122.97700574468506,
            "resetPriceTime": 1545530400000,
            "resetState": "UpwardReset",
            "state": "Trading",
            "totalSupplyA": 506.61953689367846,
            "totalSupplyB": 1013.2390737873569,
            "totalUsers": 18,
          },
        },
      }
    }
    ethBalance={
      Object {
        "allowance": 1.157920892373162,
        "eth": 0.5076699707487066,
        "weth": 0.984992,
      }
    }
    ethPrice={123}
    handleClose={[Function]}
    mozartList={
      Array [
        "0x00be45fe5903ab1b33a9d3969b05b29552a6d18b",
      ]
    }
    notify={[MockFunction]}
    visible={true}
  />
</div>
`;
